net: coap: Fix underlying type for block number

rlubos opened the pull request:
The block number in block1/2 options can be encoded on up to 20 bits according to RFC 7959, therefore the underlying type used in helper functions to retrieve the block number should be large enough to hold the result. Therefore, replace the container for block number with uint32_t instead of uint8_t.

Fixes #76000

TODO:
- [x] Add a note in the migration guide, when available.

rlubos committed:
net: coap: Fix underlying type for block number

The block number in block1/2 options can be encoded on up to 20 bits
according to RFC 7959, therefore the underlying type used in helper
functions to retrieve the block number should be large enough to hold
the result. Therefore, replace the container for block number with
uint32_t instead of uint8_t.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

pdgendt commented:
This is a public API change, we should at least have a migration note.

rlubos commented:
> This is a public API change, we should at least have a migration note.

True, although there's no migration guide for 4.0 yet. I'll add DNM and leave a TODO in the PR not to forget it.

SeppoTakalo reviewed (dismissed):
Looks OK.
Approved once the migration guide is added.

rlubos committed:
net: coap: Parse more flag in coap_get_block2_option()

Parse the more flag in coap_get_block2_option(), so that the function
can be used not only with requests but also with replies (where the more
flag should not be ignored).

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos commented:
Added a second commit which adds parsing of the more flag in coap_get_block2_option(). Still waiting for the migration guide.

rlubos committed:
doc: migration-guide-4.0: Add entry about CoAP block API change

Add entry about the coap_get_block1_option() and
coap_get_block2_option() API change.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos commented:
Added the migration guide entry for the API change.
